File: app/src/main/java/de/exercicse/jrossbach/podcast/search/PodcastListFragment.java
package de.exercicse.jrossbach.podcast.search;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;

import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import de.exercicse.jrossbach.podcast.MainActivity;
import de.exercicse.jrossbach.podcast.R;
import de.exercicse.jrossbach.podcast.player.AudioPlayerFragment;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;


public class PodcastListFragment extends Fragment implements PodcastChannelView {

    @BindView(R.id.podcast_recycler_view)
    RecyclerView recyclerView;
    PodcastItemAdapter adapter;
    private static String PODCAST_URL;

    @BindView(R.id.audio_progress_bar)
    ProgressBar progressBar;
    private SearchListPresenter presenter;

    public static PodcastListFragment newInstance(String url){
        PodcastListFragment fragment = new PodcastListFragment();
        Bundle args = new Bundle();
        args.putString("podcastUrl", url);
        fragment.setArguments(args);
        return fragment;
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        presenter = new SearchListPresenter();
        View rootView = inflater.inflate(R.layout.podcast_list_fragment, container, false);
        ButterKnife.bind(this, rootView);
        PODCAST_URL = getArguments().getString("podcastUrl");
        return rootView;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        Bundle bundle = getArguments();
        if(bundle != null){
            PODCAST_URL = getArguments().getString("podcastUrl");
        }
        initRecyclerView();
    }

    @Override
    public void onStart() {
        super.onStart();
        presenter.attachView(this);
        presenter.search(PODCAST_URL);
    }

    private void initRecyclerView() {
        adapter = new PodcastItemAdapter(this);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerView.setAdapter(adapter);
    }

    @Override
    public void onStop() {
        presenter.detachView();
        super.onStop();
    }

    @Override
    public void onItemsLoadedSuccessfully(List<PodcastItemViewModel> podcastItemViewModels) {
        adapter.setItems(podcastItemViewModels);
        adapter.notifyDataSetChanged();
    }

    @Override
    public void showProgress(boolean show) {
        progressBar.setVisibility(show ? VISIBLE : GONE);
    }

    @Override
    public void onItemClick(PodcastItemViewModel podcastItemViewModel) {
        AudioPlayerFragment fragment = AudioPlayerFragment.newInstance();
        Bundle args = new Bundle();
        args.putParcelable("podcastItemViewModel", podcastItemViewModel);
        fragment.setArguments(args);
        ((MainActivity)requireActivity()).replaceCurrentFragment(fragment);
    }

}

File: app/src/main/java/de/exercicse/jrossbach/podcaster/feature/search/ui/PodcastListFragment.kt
package de.exercicse.jrossbach.podcaster.feature.search.ui

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import de.exercicse.jrossbach.podcaster.R
import de.exercicse.jrossbach.podcaster.base.ui.BaseFragment
import de.exercicse.jrossbach.podcaster.feature.channel.ui.PodcastChannelFragment
import de.exercicse.jrossbach.podcaster.feature.player.ui.AudioPlayerFragment
import de.exercicse.jrossbach.podcaster.feature.search.ui.adapter.PodcastItemAdapter
import de.exercicse.jrossbach.podcaster.ui.MainActivity
import kotlinx.android.synthetic.main.podcast_list_fragment.*
import javax.inject.Inject

class PodcastListFragment : BaseFragment(), PodcastChannelView, PodCastItemClickListener {

    private lateinit var podcastItemAdapter: PodcastItemAdapter

    private lateinit var podCastUrl: String

    @Inject
    lateinit var presenter: SearchListPresenter

    override fun layoutId(): Int = R.layout.podcast_list_fragment

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        podCastUrl = arguments?.getString(INTENT_EXTRA_PODCAST_URL) ?: ""
        initRecyclerView()
    }

    override fun onStart() {
        super.onStart()
        presenter.attachView(this)
        presenter.search(podCastUrl)
    }

    private fun initRecyclerView() {
        podcastItemAdapter = PodcastItemAdapter(this)
        podcast_recycler_view.apply {
            layoutManager = LinearLayoutManager(context)
            adapter = podcastItemAdapter
        }
    }

    override fun onStop() {
        presenter.detachView()
        super.onStop()
    }

    override fun onItemsLoadedSuccessfully(podcastItemViewModels: List<PodcastItemViewModel>) {
        podcastItemAdapter.items = podcastItemViewModels
    }

    override fun showProgress(show: Boolean) {
        audio_progress_bar.visibility = if (show) View.VISIBLE else View.GONE
    }

    override fun onItemClick(podcastItemViewModel: PodcastItemViewModel) {
        childFragmentManager.beginTransaction()
            .add(R.id.podcast_list_container, AudioPlayerFragment.newInstance(podcastItemViewModel))
            .addToBackStack(null)
            .commit()
    }

    override fun showError(message: String) {
        error_view.apply {
            visibility = View.VISIBLE
            text = message
        }
    }

    override fun onChannelItemClicked(id: String) {
        (requireActivity() as MainActivity)
            .addFragment(PodcastChannelFragment.newInstance(id))
    }

    companion object {

        private const val INTENT_EXTRA_PODCAST_URL = "intent_extra_podcast_url"

        fun newInstance(url: String): PodcastListFragment {
            val fragment = PodcastListFragment()
            val args = Bundle()
            args.putString(INTENT_EXTRA_PODCAST_URL, url)
            fragment.arguments = args
            return fragment
        }
    }
}